controller.py

```diff
@@ -192,41 +193,96 @@ def search(self, ip_address=None, live=True):
             )
             return Response(body, 200, headers)
 
+    @classmethod
+    def opds_response_links(cls, response, rel):
+        """Find all the links in the given response for the given 
+        link relation.
+        """
+        # Look in the response itself for a Link header.
+        links = []
+        link = response.links.get(rel)
+        if link:
+            links.append(link.get('url'))
+        media_type = response.headers.get('Content-Type')
+        if media_type == OPDSCatalog.OPDS_TYPE:
+            # Parse as OPDS 2.
+            catalog = json.loads(response.content)
+            links = []
+            for k,v in catalog.get("links", {}).iteritems():
+                if k == rel:
+                    links.append(v.get("href"))
+        elif media_type == OPDSCatalog.OPDS_1_TYPE:
+            # Parse as OPDS 1.
+            feed = feedparser.parse(response.content)
+            for link in feed.get("feed", {}).get("links", []):
+                if link.get('rel') == rel:
+                    links.append(link.get('href'))
+        return [urljoin(response.url, url) for url in links if url]
+
+    @classmethod
+    def opds_response_links_to_auth_document(cls, opds_response, auth_url):
+        """Verify that the given response links to the given URL as its
+        Authentication For OPDS document.
+        
+        The link might happen in the `Link` header or in the body of
+        an OPDS feed.
+        """
+        links = []
+        try:
+            links = cls.opds_response_links(
+                opds_response, AuthenticationDocument.AUTHENTICATION_DOCUMENT_REL
+            )
+        except ValueError, e:
+            # The response itself is malformed.
+            return False
+        return auth_url in links
+
     def register(self, do_get=HTTP.get_with_timeout):
         auth_url = flask.request.form.get("url")
         if not auth_url:
             return NO_AUTH_URL
 
-        try:
-            auth_response = do_get(
-                auth_url, allowed_response_codes=["2xx", "3xx", 404],
-                timeout=30
-            )
-            # We only allowed 404 above so that we could return a more
-            # specific problem detail document if it happened.
-            if auth_response.status_code == 404:
-                return AUTH_DOCUMENT_NOT_FOUND
-        except RequestTimedOut, e:
-            logging.error(
-                "Registration of %s failed: timed out retrieving authentication document",
-                auth_url, exc_info=e
-            )
-            return AUTH_DOCUMENT_TIMEOUT
-        except Exception, e:
-            logging.error(
-                "Registration of %s failed: error retrieving authentication document", 
-                exc_info=e
-            )
-            return ERROR_RETRIEVING_AUTH_DOCUMENT
+        def _make_request(url, on_404, on_timeout, on_exception):
```

Comment:
This method is used to get the A4OPDS document and the OPDS feed that it links to. If there's a problem, a generic problem detail document is detailed with more specific errors.